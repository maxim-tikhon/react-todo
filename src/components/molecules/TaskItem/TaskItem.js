import ClearIcon from '@mui/icons-material/Clear';
import CheckBoxInput from '../../atoms/CheckBoxInput/CheckBoxInput';
import styles from './TaskItem.module.scss';

function TaskItem({ task, onDeleteTask, onCheckTask }) {
  return (
    <>
      <CheckBoxInput
        label={task.description}
        checked={task.completed}
        onCheck={(checked) => onCheckTask({ id: task.id, checked })}
      />
      <ClearIcon className={styles.deleteIcon} onClick={() => onDeleteTask(task.id)} />
    </>
  );
}

export default TaskItem;
