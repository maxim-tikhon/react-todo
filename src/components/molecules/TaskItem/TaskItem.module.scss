.deleteIcon {
  margin-left: auto;
  cursor: pointer;
  color: grey;

  &:hover {
    color: black;
  }
}
