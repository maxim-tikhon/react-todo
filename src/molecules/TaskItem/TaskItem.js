import CheckBoxInput from "../../atoms/CheckBoxInput/CheckBoxInput";
import ClearIcon from "@mui/icons-material/Clear";
import styles from "./TaskItem.module.scss";

const TaskItem = ({ task }) => {
	const taskClasses = task.completed
		? `${styles.task} ${styles.completed}`
		: styles.task;

	return (
		<div>
			{/* doesn't work because state is not updated */}
			<CheckBoxInput checked={task.completed} />
			<span className={taskClasses}>{task.description}</span>
			<ClearIcon className={styles.deleteIcon} />
		</div>
	);
};

export default TaskItem;
