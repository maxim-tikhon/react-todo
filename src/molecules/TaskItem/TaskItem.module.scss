.task {
	max-width: calc(100% - 4rem);
	white-space: nowrap;
	overflow: hidden;
	text-overflow: ellipsis;
	margin-left: 10px;
	cursor: pointer;
	outline: none;

	&.completed {
		text-decoration: line-through;
		color: lightgray;
	}
}

.deleteIcon {
	margin-left: auto;
	cursor: pointer;
	color: grey;

	&:hover {
		color: black;
	}
}
