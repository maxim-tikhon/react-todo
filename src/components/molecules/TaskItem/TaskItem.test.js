import renderer from 'react-test-renderer';
import { fireEvent, render, screen } from '@testing-library/react';
import TaskItem from './TaskItem';

describe('TaskItem', () => {
  const mockTask = {
    id: 1,
    description: 'Task description',
    completed: false,
  };

  const mockOnDeleteTask = jest.fn();
  const mockOnCheckTask = jest.fn();

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const task = {
      id: 1,
      description: 'Test Task',
      completed: true,
    };

    const output = renderer.create(<TaskItem task={task} />).toJSON();
    expect(output).toMatchSnapshot();
  });

  it('renders task description', () => {
    render(<TaskItem task={mockTask} />);

    const taskDescription = screen.getByText('Task description');

    expect(taskDescription).toBeInTheDocument();
  });

  test('renders unchecked checkbox and active task when task is not completed', () => {
    const task = { ...mockTask };

    render(<TaskItem task={task} />);

    const checkbox = screen.getByRole('checkbox');

    expect(checkbox.checked).toBe(false);
  });

  test('renders checked checkbox and completed task when task is completed', () => {
    const completedTask = { ...mockTask, completed: true };

    render(<TaskItem task={completedTask} />);

    const checkbox = screen.getByRole('checkbox');

    expect(checkbox.checked).toBe(true);
  });

  test('calls onDeleteTask callback when ClearIcon is clicked', () => {
    const task = { ...mockTask };

    render(<TaskItem task={task} onDeleteTask={mockOnDeleteTask} onCheckTask={mockOnCheckTask} />);

    const clearIcon = screen.getByTestId('ClearIcon');

    fireEvent.click(clearIcon);

    expect(mockOnDeleteTask).toHaveBeenCalledTimes(1);
    expect(mockOnDeleteTask).toHaveBeenCalledWith(1);
  });

  test('calls onCheckTask callback when task is clicked', () => {
    const task = { ...mockTask };

    render(<TaskItem task={task} onDeleteTask={mockOnDeleteTask} onCheckTask={mockOnCheckTask} />);

    const taskDescription = screen.getByText('Task description');

    fireEvent.click(taskDescription);

    expect(mockOnCheckTask).toHaveBeenCalledTimes(1);
    expect(mockOnCheckTask).toHaveBeenCalledWith({ id: 1, checked: true });
  });

  test('calls onCheckTask callback when checkbox is clicked', () => {
    const task = { ...mockTask };

    render(<TaskItem task={task} onDeleteTask={mockOnDeleteTask} onCheckTask={mockOnCheckTask} />);

    const checkbox = screen.getByRole('checkbox');

    fireEvent.click(checkbox);

    expect(mockOnCheckTask).toHaveBeenCalledTimes(1);
    expect(mockOnCheckTask).toHaveBeenCalledWith({ id: 1, checked: true });
  });
});
