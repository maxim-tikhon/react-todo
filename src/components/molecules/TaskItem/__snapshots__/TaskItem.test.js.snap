// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskItem renders correctly 1`] = `
Array [
  <input
    checked={true}
    className="customCheckbox"
    id="Test Task"
    onChange={[Function]}
    type="checkbox"
  />,
  <label
    className="label"
    htmlFor="Test Task"
  >
    Test Task
  </label>,
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium deleteIcon css-i4bv87-MuiSvgIcon-root"
    data-testid="ClearIcon"
    focusable="false"
    onClick={[Function]}
    viewBox="0 0 24 24"
  >
    <path
      d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
    />
  </svg>,
]
`;
